# Финальный спринт. Задачи по SQL.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-и-задачи" data-toc-modified-id="Цели-и-задачи-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели и задачи</a></span><ul class="toc-item"><li><span><a href="#Задания" data-toc-modified-id="Задания-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задания</a></span></li></ul></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Книги-books" data-toc-modified-id="Книги-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Книги <code>books</code></a></span></li><li><span><a href="#Авторы-authors" data-toc-modified-id="Авторы-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Авторы <code>authors</code></a></span></li><li><span><a href="#Издательства-publishers" data-toc-modified-id="Издательства-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательства <code>publishers</code></a></span></li><li><span><a href="#Оценки-книг-пользователями-ratings" data-toc-modified-id="Оценки-книг-пользователями-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Оценки книг пользователями <code>ratings</code></a></span></li><li><span><a href="#Обзоры-книг-пользователями-reviews" data-toc-modified-id="Обзоры-книг-пользователями-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обзоры книг пользователями <code>reviews</code></a></span></li></ul></li><li><span><a href="#Подключение-к-БД.-Обзор-таблиц." data-toc-modified-id="Подключение-к-БД.-Обзор-таблиц.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подключение к БД. Обзор таблиц.</a></span></li><li><span><a href="#Решения-задач" data-toc-modified-id="Решения-задач-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Решения задач</a></span><ul class="toc-item"><li><span><a href="#Сколько-вышло-книг-после-1-января-2000-года" data-toc-modified-id="Сколько-вышло-книг-после-1-января-2000-года-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Сколько вышло книг после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-каждой-книги-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка каждой книги</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-больше-всего-книг-(толще-50-страниц)" data-toc-modified-id="Определить-издательство,-которое-выпустило-больше-всего-книг-(толще-50-страниц)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило больше всего книг (толще 50 страниц)</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-(книги-только-с-50-и-более-оценками)" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-(книги-только-с-50-и-более-оценками)-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг (книги только с 50 и более оценками)</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили более 50 оценок</a></span></li></ul></li><li><span><a href="#Общий-вывод-исследования" data-toc-modified-id="Общий-вывод-исследования-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод исследования</a></span></li></ul></div>

## Цели и задачи

Задача - анализ базы данных крупного сервиса для чтения книг по подписке. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


## Описание данных

![](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2)


### Книги `books`

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

### Авторы `authors`

- `author_id` — идентификатор автора;
- `author` — имя автора.

### Издательства `publishers`

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

### Оценки книг пользователями `ratings`

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

### Обзоры книг пользователями `reviews`

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Подключение к БД. Обзор таблиц.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
def headline(line, width=79):
    """Return the string in plain uppercase text with pseudographics"""
    return f'\n{line.upper():=^{width}}'

In [3]:
def get(query):
    return pd.io.sql.read_sql(query, con = engine)

In [4]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [5]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = f'SELECT * FROM {table} LIMIT 5;'
    print(headline(table))
    display(get(query))


=====================================BOOKS=====================================


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


====================================AUTHORS====================================


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


==================================PUBLISHERS===================================


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


====================================RATINGS====================================


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


====================================REVIEWS====================================


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Выводы**:

- Реальная структура базы данных совпадает с описанной.

## Решения задач

### Сколько вышло книг после 1 января 2000 года

In [6]:
query = """
SELECT COUNT(*)
FROM books
WHERE publication_date > '2000-01-01';
"""

get(query)

count
0    819

**Выводы**:

- После 1 января 2000 выпущено 819 книг.
- Строго говоря, не только книг, а любых изданий - поскольку в третьей задачей книгой называют конкретно издание толще 50 страниц.
- Строго говоря, сама дата 1 января 2000 года в выборку не входит, поскольку так сформулирована задача.

In [7]:
query = """
SELECT COUNT(*)
FROM books
WHERE publication_date >= '2000-01-01';
"""

get(query)

count
0    821

- Если поставить вопрос чуть иначе, то начиная с 1 января 2000 года вышло 821 издание.

### Количество обзоров и средняя оценка каждой книги

In [8]:
query = """
SELECT author,
       title,
       publication_date,
       COALESCE(reviews_count, 0) AS reviews_count,
       COALESCE(avg_rating, 0) AS avg_rating
FROM books AS b
LEFT JOIN authors AS a
    ON b.author_id = a.author_id
LEFT JOIN (SELECT book_id, COUNT(*) AS reviews_count
           FROM reviews
           GROUP BY book_id) AS rv
           ON b.book_id = rv.book_id
LEFT JOIN (SELECT book_id, AVG(rating) AS avg_rating
           FROM ratings
           GROUP BY book_id) AS ra
           ON b.book_id = ra.book_id
ORDER BY reviews_count DESC, avg_rating DESC;
"""

get(query)

author  \
0                                      Stephenie Meyer   
1                           J.K. Rowling/Mary GrandPré   
2                           J.K. Rowling/Mary GrandPré   
3                    Markus Zusak/Cao Xuân Việt Khương   
4                                      Jeannette Walls   
..                                                 ...   
995                      Leonardo da Vinci/H. Anna Suh   
996                 Teddy Slater/Ron Dias/Ric González   
997                 Edgar Allan Poe/Benjamin F. Fisher   
998  Faye Perozich/Anne Rice/John Bolton/Daerick Gröss   
999                      Kimon Nicolaides/Mamie Harmon   

                                                 title publication_date  \
0                              Twilight (Twilight  #1)       2006-09-06   
1    Harry Potter and the Prisoner of Azkaban (Harr...       2004-05-01   
2    Harry Potter and the Chamber of Secrets (Harry...       1999-06-02   
3                                       The Book Thief       2006-03-14   
4                                     The Glass Castle       2006-01-17   
..                                                 ...              ...   
995                               Leonardo's Notebooks       2005-08-01   
996  Disney's Beauty and the Beast (A Little Golden...       2004-05-11   
997                          Essential Tales and Poems       2004-10-25   
998    Anne Rice's The Vampire Lestat: A Graphic Novel       1991-11-30   
999                            The Natural Way to Draw       1990-02-01   

     reviews_count  avg_rating  
0                7    3.662500  
1                6    4.414634  
2                6    4.287500  
3                6    4.264151  
4                6    4.206897  
..             ...         ...  
995              0    4.000000  
996              0    4.000000  
997              0    4.000000  
998              0    3.666667  
999              0    3.000000  

[1000 rows x 5 columns]

**Выводы**:

- Получена таблица, отсортированная по количеству обзоров и среднему рейтингу.
- Нам доступна ровно тысяча записей. Это совпадение, или ограничение?
- Самые обсуждаемые книги написаны женщинами.

### Определить издательство, которое выпустило больше всего книг (толще 50 страниц)

In [9]:
query = """
SELECT publisher, book_count
FROM publishers AS p
JOIN (SELECT COUNT(*) AS book_count,
             publisher_id
      FROM books
      WHERE num_pages > 50
      GROUP BY publisher_id
      ORDER BY book_count DESC
      LIMIT 1) AS b
      ON p.publisher_id = b.publisher_id
"""

get(query)

publisher  book_count
0  Penguin Books          42

**Выводы**:

- Больше всего полноценных книг выпустило издательство "Penguin Books" - 42 таких издания.

### Определить автора с самой высокой средней оценкой книг (книги только с 50 и более оценками)

In [10]:
query = """
SELECT author,
       author_avg_rating
FROM authors AS a
JOIN (SELECT author_id,
             AVG(book_avg_rating) AS author_avg_rating
      FROM books as b
      JOIN (SELECT book_id,
                   AVG(rating) AS book_avg_rating
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(*) > 50) AS r
            ON r.book_id = b.book_id
      GROUP BY author_id
      ORDER BY author_avg_rating DESC
      LIMIT 1) AS bs
      ON bs.author_id = a.author_id
"""

get(query)

author  author_avg_rating
0  J.K. Rowling/Mary GrandPré           4.283844

**Выводы**:

- Наиболее высокая средняя оценка книг - у дуэта авторов "J.K. Rowling/Mary GrandPré".

### Среднее количество обзоров от пользователей, которые поставили более 50 оценок

Первый же вариант, который пришёл на ум:

In [11]:
query = """
SELECT 1.0 * COUNT(review_id) / COUNT(DISTINCT username) AS avg_reviews
from reviews AS rv
WHERE username IN (SELECT username
                   FROM ratings
                   GROUP BY username
                   HAVING COUNT(rating) > 50)
"""

get(query)

avg_reviews
0    24.333333

На всякий случай запишу и второй вариант:

In [12]:
query = """
SELECT AVG(reviews_count) AS avg_reviews
FROM (SELECT username,
             COUNT(review_id) AS reviews_count
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating) > 50)
      GROUP BY username) AS r
"""

get(query)

avg_reviews
0    24.333333

**Выводы**:

- Пользователи, которые поставили более 50 оценок, в среднем публиковали 24+1/3 обзоров.

## Общий вывод исследования

- Книги Джоан Роулинг - самый обсуждаемый и ценимый товар.